File: code/matlab/readCorinet.m
function readCorinet(fileName,createNames,execution)
%  Reads result files in corinet markup language file format
%   "readCorinet(fileName,createNames,execution)
%
%   "fileName" contains the path and name of the file to be read
%   "createNames" - optional - if 1, named variables are created in the
%   calling function; if 0, one cell array is created
%   "execution" - optional - a number; if specified, only the data 
%   associated with that execution are read

fid= fopen(fileName,'rt');

if (fid<0) 
    fprintf(2,'Error opening file %s. Does it exist? Are you sure?\n',fileName);
    return;
end

if (nargin == 1)
    createNames = 0;
end

token = fscanf(fid,'%c',1);
while (~feof(fid))
    if (strcmp(token,'<'))
        token = fscanf(fid,'%c',1);
        if (strcmp(token,'?')) %scan through the prolog
            scanUntil(fid,'>');
        elseif (strcmp(token,'!')) %scan trough a comment
            token = fscanf(fid,'%c',2);
            if (strcmp(token,'--')) %start of a comment, scan until the end
                scanUntil(fid,'-->');
            else
                scanUntil(fid,'>'); %start of a processing instruction, scan until the end
            end    
        elseif (strcmp(token,'/')) %scan through a closing tag
            tagName = scanTag(fid);
            switch (tagName)
                case 'v'
                    if (synapses)
                        vCounter = vCounter + 1;
                    end
                case 'w'
                    vCounter = vCounter + 1;
                case 'p'
                    vCounter = vCounter + 1;
                case 'cycle'
                    assignin('caller','tempLabels',thisCycleTimeLabels);
                    assignin('caller','tempResults',thisCycleResults);
                    if (createNames)
                        evalin('caller',sprintf('%s{%d}{1} = %s;',varName,cycleCounter,'tempLabels'));
                        evalin('caller',sprintf('%s{%d}{2} = %s;',varName,cycleCounter,'tempResults'));
                    else
                        if (netCounter) 
                            evalin('caller',sprintf('results{%d}{1}{%d}{%d}{%d}{1} = %s;',exCounter,netCounter,elCounter,cycleCounter,'tempLabels'));
                            evalin('caller',sprintf('results{%d}{1}{%d}{%d}{%d}{2} = %s;',exCounter,netCounter,elCounter,cycleCounter,'tempResults'));
                        else
                            evalin('caller',sprintf('results{%d}{2}{%d}{%d}{1} = %s;',exCounter,taskCounter,cycleCounter,'tempLabels'));
                            evalin('caller',sprintf('results{%d}{2}{%d}{%d}{2} = %s;',exCounter,taskCounter,cycleCounter,'tempResults'));
                        end
                    end
                    cycleCounter = cycleCounter + 1;
                case 'values'
                    elCounter = elCounter + 1;
                case 'weights'
                    elCounter = elCounter + 1;
                case 'synapses'
                    elCounter = elCounter + 1;
                case 'network'
                    netCounter = netCounter + 1;
                case 'task'
                    taskCounter = taskCounter + 1;
                case 'execution'
                    exCounter = exCounter + 1;
            end
        else %scan through an opening tag
            tagName = [token scanTag(fid)];
            switch (tagName)
                case 'annotation'
                    scanUntil(fid,'</annotation>');
                case 'results'
                    scanUntil(fid,'>');
                    exCounter = 1;
                    if (~createNames)
                        assignin('caller','results',cell(0));
                    end
                case 'execution'
                    if (~createNames)
                        evalin('caller',sprintf('results{%d} = cell(1,2);',exCounter));
                    end
                    if (nargin == 3) 
                        if (exCounter < execution) %skip this entire execution
                            scanUntil(fid,'</execution>');
                            exCounter = exCounter + 1;
                        elseif (exCounter > execution) %stop reading
                            break;
                        end
                    end
                    netCounter = 1;
                    taskCounter = 1;
                case 'network'
                    %scan the rest of this tag to see if it contains data
                    tagRemainder = scanUntilAndReturn(fid,'>');
                    if (isempty(strfind(tagRemainder,'/')))
                        if (~createNames)
                            evalin('caller',sprintf('results{%d}{1}{%d} = cell(0);',exCounter,netCounter));
                        end
                        elCounter = 1;
                    end
                case 'task'
                    netCounter = 0;
                    %scan the rest of this tag to see if it contains data
                    tagRemainder = scanUntilAndReturn(fid,'>');
                    if (isempty(strfind(tagRemainder,'/')))
                        idName = getAttrVal('id',tagRemainder);
                        if (createNames)
                            varName = sprintf('taskPatterns_%s_Ex%d',idName,exCounter);
                            assignin('caller',varName,cell(0));
                        else
                            evalin('caller',sprintf('results{%d}{2}{%d} = cell(0);',exCounter,taskCounter));
                        end
                        cycleCounter = 1;
                        numVals = 0;
                    end
                case 'values'
                    tagRemainder = scanUntilAndReturn(fid,'>');
                    if (isempty(strfind(tagRemainder,'/')))
                        idName = getAttrVal('id',tagRemainder);
                        typeName = getAttrVal('type',tagRemainder);
                        if (createNames)
                            varName = sprintf('%s_%s_Ex%d',typeName,idName,exCounter);
                            assignin('caller',varName,cell(0));
                        else
                            evalin('caller',sprintf('results{%d}{1}{%d}{%d} = cell(0);',exCounter,netCounter,elCounter));
                        end
                        cycleCounter = 1;
                        numVals = 0;
                    end
                    synapses = 0;
                case 'weights'
                    tagRemainder = scanUntilAndReturn(fid,'>');
                    if (isempty(strfind(tagRemainder,'/')))
                        idName = getAttrVal('id',tagRemainder);
                        if (createNames)
                            varName = sprintf('weights_%s_Ex%d',idName,exCounter);
                            assignin('caller',varName,cell(0));
                        else
                            evalin('caller',sprintf('results{%d}{1}{%d}{%d} = cell(0);',exCounter,netCounter,elCounter));
                        end
                        cycleCounter = 1;
                        numVals = 0;
                    end
                case 'synapses'
                    tagRemainder = scanUntilAndReturn(fid,'>');
                    if (isempty(strfind(tagRemainder,'/')))
                        idName = getAttrVal('id',tagRemainder);
                        if (createNames)
                            varName = sprintf('synapses_%s_Ex%d',idName,exCounter);
                            assignin('caller',varName,cell(0));
                        else
                            evalin('caller',sprintf('results{%d}{1}{%d}{%d} = cell(0);',exCounter,netCounter,elCounter));
                        end
                        cycleCounter = 1;
                        numVals = 0;
                    end
                    synapses = 1;
                case 'cycle'
                    tagRemainder = scanUntilAndReturn(fid,'>');
                    if (createNames)
                        evalin('caller',sprintf('%s{%d} = cell(1,2);',varName,cycleCounter));
                    else
                        if (netCounter)
                            evalin('caller',sprintf('results{%d}{1}{%d}{%d}{%d} = cell(1,2);',exCounter,netCounter,elCounter,cycleCounter));
                        else
                            evalin('caller',sprintf('results{%d}{2}{%d}{%d} = cell(1,2);',exCounter,taskCounter,cycleCounter));
                        end
                    end
                    thisCycleResults = [];
                    thisCycleTimeLabels= [];
                    vCounter = 1;   
                case 'w'
                    tagRemainder = scanUntilAndReturn(fid,'>');
                    iCounter = str2num(getAttrVal('i',tagRemainder)) + 1;
                    if (~isempty(findstr(tagRemainder,'s')))
                        sCounter = str2num(getAttrVal('s',tagRemainder)) + 1;
                    else
                        sCounter = 0;
                    end
                    thisCycleTimeLabels(vCounter,:) = [iCounter sCounter];
                    subCounter = 1;
                case 'p'
                    tagRemainder = scanUntilAndReturn(fid,'>');
                    iCounter = str2num(getAttrVal('i',tagRemainder)) + 1;
                    if (~isempty(findstr(tagRemainder,'s')))
                        sCounter = str2num(getAttrVal('s',tagRemainder)) + 1;
                    else
                        sCounter = 0;
                    end
                    thisCycleTimeLabels(vCounter,:) = [iCounter sCounter];
                    subCounter = 1;
                case 'v'
                    tagRemainder = scanUntilAndReturn(fid,'>');
                    iCounter = str2num(getAttrVal('i',tagRemainder)) + 1;
                    if (~isempty(findstr(tagRemainder,'s')))
                        sCounter = str2num(getAttrVal('s',tagRemainder)) + 1;
                    else
                        sCounter = 0;
                    end
                    thisCycleTimeLabels(vCounter,:) = [iCounter sCounter];
                    if (~synapses)
                        if (numVals) 
                            thisCycleResults(vCounter,:) = fscanf(fid,'%g',numVals)';
                        else
                            valuesStr = scanUntilAndReturn(fid,'<');
                            temp = str2num(valuesStr(1:end-1));
                            numVals = length(temp);
                            thisCycleResults(vCounter,:) = temp;
                        end
                        vCounter = vCounter + 1;
                    else
                        subCounter = 1;
                    end
                case 'n'
                    if (numVals) 
                        thisCycleResults(:,subCounter,vCounter) = fscanf(fid,'%g',numVals);
                    else
                        valuesStr = scanUntilAndReturn(fid,'<');
                        temp = str2num(valuesStr(1:end-1));
                        numVals = length(temp);
                        thisCycleResults(:,subCounter,vCounter) = temp;
                    end     
                    subCounter = subCounter + 1;
                case 'r'
                    if (numVals) 
                        thisCycleResults(subCounter,:,vCounter) = fscanf(fid,'%g',numVals)';
                    else
                        valuesStr = scanUntilAndReturn(fid,'<');
                        temp = str2num(valuesStr(1:end-1));
                        numVals = length(temp);
                        thisCycleResults(subCounter,:,vCounter) = temp;
                    end  
                    subCounter = subCounter + 1;
            end
        end
    end
    token = fscanf(fid,'%c',1);
end
evalin('caller','clear tempResults;');
evalin('caller','clear tempLabels;');
fclose(fid);

%---------------------------------------------------------------------

function attrVal = getAttrVal(attrName,str)
%  Get the attribute value
attrStart = strfind(str,attrName);
attrVal = strtok(str(attrStart+length(attrName)+2:end),'"');

function tagName = scanTag(fid)
%  Scan the tag name
tagName = '';
token = '';
while (~strcmp(token,' ') & ~strcmp(token,'>'))
    tagName = [tagName token];
    token = fscanf(fid,'%c',1);
end

function scanUntil(fid,str)
%  Scans until it encounters the string 'str'
token = '';
while (1)
    while (~strcmp(token,str(1)))
        token = fscanf(fid,'%c',1);
    end
    if (length(str) > 1)
        token = [token fscanf(fid,'%c',length(str)-1)];
    end
    if (strcmp(token,str))
        break;
    else
        fseek(fid,-length(str)+1,0);
    end
end

function returnStr = scanUntilAndReturn(fid,str)
%  Scans until it encounters the string 'str' and returns the scanned
%  string.
token = '';
returnStr = '';
while (1)
    while (~strcmp(token,str(1)))
        returnStr = [returnStr token];
        token = fscanf(fid,'%c',1);
    end
    if (length(str) > 1)
        token = [token fscanf(fid,'%c',length(str)-1)];
    end
    if (strcmp(token,str))
        returnStr = [returnStr token];
        break;
    else
        returnStr = [returnStr token(1)];
        fseek(fid,-length(str)+1,0);
    end
end


    

File: code/r/readCorinet.R
readCorinet <-function(fileName,createNames = FALSE,execution = 0)
#  Reads result files in CORINET Markup Language file format
#   "readCorinet(fileName,execution)
#
#   "fileName" contains the path and name of the file to be read
#   "createNames" - optional - if TRUE, named variables are created in the
#   calling function; if FALSE, one cell array is created
#   "execution" if > 0 only the data 
#   associated with that execution are read

{
	helperPath <- paste(Sys.getenv("CORINETROOT"),"code/r/corinetHelperFunctions.R",sep="")
	source(helperPath)

	fid <- file(description = fileName,open = "rt")
	caller = parent.frame()
    
	token <- readChar(fid,1)
	while (length(token)) {
        if (identical(token,"<")) {
        	token <- readChar(fid,1)
        	if (identical(token,"?")) {#scan through the prolog
           		scanUntil(fid,">")
			}
       	    else if (identical(token,'!')) {#scan through a comment
           	    token <- readChar(fid,2)
           	    if (identical(token,'--')) {#start of a comment, scan until the end
               	    scanUntil(fid,'-->')
			    }
                else {#start of a processing instruction, scan until the end
                    scanUntil(fid,'>') 
			    }
		    }
		    else if (identical(token,'/')) {#scan through a closing tag
          	    tagName = scanTag(fid)
          	    if (identical(tagName,'results')) {
          	        eval(expression(rm(exName,elemName,cycleName)),envir=caller)
				    break
				} 
				else if (identical(tagName,'execution')) {
				    exCounter <- exCounter + 1
		        }
				else if (identical(tagName,'cycle')) {
				    assign('tempLabels',matrix(thisCycleTimeLabels,ncol=2,byrow=TRUE,dimnames=list(c(),c('iteration','step'))),caller)
				    assign('tempValues',thisCycleValues,caller)
				    eval(expression(results[[exName]][[elemName]][[cycleName]]$time <- tempLabels),envir=caller)
				    eval(expression(results[[exName]][[elemName]][[cycleName]]$values <- tempValues),envir=caller)
				    eval(expression(rm(tempLabels,tempValues)),envir=caller)
				    cycleCounter <- cycleCounter + 1
				}
				else if (identical(tagName,'v') && noSubs) {
				    thisCycleValues[[vCounter]] <- tempVals
				    vCounter <- vCounter + 1
				}
				else if ((identical(tagName,'v') && !noSubs) || identical(tagName,'w')) {
				    thisCycleValues[[vCounter]] <- matrix(data=tempVals,ncol=subCounter,byrow=FALSE)
				    vCounter <- vCounter + 1
				}
				else if (identical(tagName,'p')) {
				    thisCycleValues[[vCounter]] <- matrix(data=tempVals,nrow=subCounter,byrow=TRUE)
				    vCounter <- vCounter + 1
				}
			}
			else {#scan through an opening tag
			    tagName <- append(token,scanTag(fid))
           	    tagName <- paste(tagName,sep="",collapse="")
				if (identical(tagName,'annotation')) {
                    scanUntil(fid,'</annotation>')
				}
                else if (identical(tagName,'results')) {
                    scanUntil(fid,'>')
                    exCounter <- 1
                    eval(expression(results <- list()),env=caller)
				}
				else if (identical(tagName,'execution')) {
				    assign('exName',sprintf('execution%d',exCounter),env=caller)
                    eval(expression(results[[exName]] <- list()),envir=caller)
                    if (execution > 0) {
                        if (exCounter < execution) {#skip this entire execution
                            scanUntil(fid,'</execution>')
                            exCounter <- exCounter + 1
                        }
                        else if (exCounter > execution) {#stop reading
                            break
                        }
                    }
				}
				else if (identical(tagName,'network')) {
                    tagRemainder <- scanUntilAndReturn(fid,'>')
				}
				else if (identical(tagName,'task') || identical(tagName,'synapses') || identical(tagName,'weights') || identical(tagName,'values')) {
                    tagRemainder <- scanUntilAndReturn(fid,'>')
                    noSubs <- FALSE
                    if (!length(grep('/',tagRemainder))) {
                        idName <- getAttrVal('id',tagRemainder)
                        if (identical(tagName,'task')) {
                            assign('elemName',sprintf('taskPatterns_%s',idName),envir=caller)
                        }
                        else if (identical(tagName,'synapses')) {
                            assign('elemName',sprintf('synapses_%s',idName),envir=caller)
                        }
                        else if (identical(tagName,'weights')) {
                            assign('elemName',sprintf('weights_%s',idName),envir=caller)
                        }
                        else if (identical(tagName,'values')) {
                            assign('elemName',sprintf('%s_%s',getAttrVal('type',tagRemainder),idName),envir=caller)
                            noSubs = TRUE
                        }
                        eval(expression(results[[exName]][[elemName]] <- list()),envir=caller)
                        cycleCounter <- 1
                        numVals <- 0
                    }
				}
				else if (identical(tagName,'cycle')) {
				    tagRemainder <- scanUntilAndReturn(fid,'>')
				    assign('cycleName',sprintf('cycle%d',cycleCounter),caller)
				    eval(expression(results[[exName]][[elemName]][[cycleName]] <- list('time'=0,'values'=0)),envir=caller)
				    thisCycleTimeLabels <- vector()
				    thisCycleValues <- list()
                    vCounter <- 1   
				}
				else if (identical(tagName,'p') || identical(tagName,'w') || identical(tagName,'v')) {
				    tagRemainder <- scanUntilAndReturn(fid,'>')
                    iteration <- type.convert(getAttrVal('i',tagRemainder))[1] + 1
                    if (length(grep('s',tagRemainder))) {
                        step <- type.convert(getAttrVal('s',tagRemainder))[1] + 1
                    }
                    else {
                        step <- 0
                    }
                    thisCycleTimeLabels <- append(thisCycleTimeLabels,c(iteration,step))
                    tempVals <- vector()
                    subCounter <- 0
				}
				if (identical(tagName,'r') || identical(tagName,'n') || (identical(tagName,'v') && noSubs)) {
				    valstr = scanUntilAndReturn(fid,"<")
                    vals = strsplit(valstr,split=" ")[[1]]
                    for(j in 1:length(vals)) {
                        if (!identical(vals[j],"")) {
                            tempVals <- append(tempVals,type.convert(vals[j]))
                   	    }
                   	}
                    subCounter <- subCounter + 1
				}
			}
	    }
	    else {
	        token <- readChar(fid,1)
	    }
    }
    close(fid)
}